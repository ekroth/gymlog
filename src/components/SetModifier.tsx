import React from 'react'
import { Button, Text, Grid, Row, Col } from 'native-base'
import { NumberInputComponent } from './NumberInput'

type SetModifierProps = {
  onSetWeight?: (weight: number) => void
  onSetReps?: (weight: number) => void

  onLeftButtonPress?: () => void
  onRightButtonPress?: () => void

  leftButtonText: string
  rightButtonText: string

  initWeight: number
  initReps: number
}

export const SetModifierComponent = (props: SetModifierProps) => (
  <Grid style={{ alignItems: 'center' }}>
    <Row>
      <NumberInputComponent
        type={'number'}
        value={props.initWeight}
        onChange={props.onSetWeight}
        step={2.5}
        minValue={0}
      />
    </Row>
    <Row>
      <NumberInputComponent
        type={'real'}
        value={props.initReps}
        onChange={props.onSetReps}
        minValue={0}
        step={1}
      />
    </Row>
    <Row>
      <Col size={1}>
        <Button
          onPress={props.onLeftButtonPress}
          style={{ alignSelf: 'center' }}
        >
          <Text>{props.leftButtonText}</Text>
        </Button>
      </Col>
      <Col size={1}>
        <Button
          onPress={props.onRightButtonPress}
          style={{ alignSelf: 'center' }}
        >
          <Text>{props.rightButtonText}</Text>
        </Button>
      </Col>
    </Row>
  </Grid>
)
