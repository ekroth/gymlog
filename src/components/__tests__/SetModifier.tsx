import React from 'react'
import { shallow } from 'enzyme'
import { SetModifierComponent } from '../SetModifier'
import {
  ExerciseModifierComponent,
  ExerciseModifierProps
} from '../ExerciseModifier'

describe('ExerciseModifier', () => {
  describe('rendering', () => {
    it('should match snapshot for zero sets', () => {
      const props: ExerciseModifierProps = {
        sets: []
      }
      const wrapper = shallow(<ExerciseModifierComponent {...props} />)
      expect(wrapper).toMatchSnapshot()
    })

    const sharedProps: ExerciseModifierProps = {
      initSet: {
        reps: 1,
        weight: 170
      },
      sets: [
        {
          reps: 5,
          weight: 100
        },
        {
          reps: 2,
          weight: 150
        }
      ]
    }

    it('should match snapshot for multiple sets', () => {
      const wrapper = shallow(<ExerciseModifierComponent {...sharedProps} />)
      expect(wrapper).toMatchSnapshot()
    })

    it('should call callback when adding a set', () => {
      const cb = jest.fn()

      const props: ExerciseModifierProps = {
        onAddSet: cb,
        ...sharedProps
      }

      const wrapper = shallow(<ExerciseModifierComponent {...props} />)
      wrapper
        .find(SetModifierComponent)
        .first()
        .props().onLeftButtonPress!()

      expect(cb.mock.calls.length).toEqual(1)
    })

    it('should be able to clear the current set weights', () => {
      const wrapper = shallow(<ExerciseModifierComponent {...sharedProps} />)
      wrapper
        .find(SetModifierComponent)
        .first()
        .props().onRightButtonPress!()

      expect(
        (wrapper.instance() as ExerciseModifierComponent).state.current
      ).toEqual({ reps: 0, weight: 0 })
    })

    it('should call callback when deleting a set', () => {
      const cb = jest.fn()

      const props: ExerciseModifierProps = {
        onDeleteSet: cb,
        selected: 1,
        ...sharedProps
      }

      const wrapper = shallow(<ExerciseModifierComponent {...props} />)
      wrapper
        .find(SetModifierComponent)
        .first()
        .props().onRightButtonPress!()

      expect(cb.mock.calls).toEqual([[props.selected]])
      expect(
        (wrapper.instance() as ExerciseModifierComponent).state.selected
      ).toBeUndefined()
    })

    it('should call callback when modifying a set', () => {
      const cb = jest.fn()

      const props: ExerciseModifierProps = {
        onModifySet: cb,
        selected: 1,
        ...sharedProps
      }

      const wrapper = shallow(<ExerciseModifierComponent {...props} />)
      wrapper
        .find(SetModifierComponent)
        .first()
        .props().onLeftButtonPress!()

      expect(cb.mock.calls).toEqual([[props.selected, props.initSet]])
      expect(
        (wrapper.instance() as ExerciseModifierComponent).state.selected
      ).toBeUndefined()
    })
  })
})
