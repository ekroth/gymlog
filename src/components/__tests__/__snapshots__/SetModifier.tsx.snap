// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseModifier rendering should match snapshot for multiple sets 1`] = `
<GridNB>
  <RowNB
    size={40}
  >
    <Component
      initReps={1}
      initWeight={170}
      leftButtonText="Add"
      onLeftButtonPress={[Function]}
      onRightButtonPress={[Function]}
      onSetReps={[Function]}
      onSetWeight={[Function]}
      rightButtonText="Clear"
    />
  </RowNB>
  <RowNB
    size={60}
  >
    <FlatList
      data={
        Array [
          Object {
            "reps": 5,
            "weight": 100,
          },
          Object {
            "reps": 2,
            "weight": 150,
          },
        ]
      }
      disableVirtualization={false}
      extraData={
        Object {
          "current": Object {
            "reps": 1,
            "weight": 170,
          },
          "selected": undefined,
        }
      }
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </RowNB>
</GridNB>
`;

exports[`ExerciseModifier rendering should match snapshot for zero sets 1`] = `
<GridNB>
  <RowNB
    size={40}
  >
    <Component
      initReps={0}
      initWeight={0}
      leftButtonText="Add"
      onLeftButtonPress={[Function]}
      onRightButtonPress={[Function]}
      onSetReps={[Function]}
      onSetWeight={[Function]}
      rightButtonText="Clear"
    />
  </RowNB>
  <RowNB
    size={60}
  >
    <FlatList
      data={Array []}
      disableVirtualization={false}
      extraData={
        Object {
          "current": Object {
            "reps": 0,
            "weight": 0,
          },
          "selected": undefined,
        }
      }
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </RowNB>
</GridNB>
`;
